### PR TITLE
Track Workers Progress

### PR DESCRIPTION
**Pull request by capella**
This pr tracks the state of each worker related to a job.
Also, export some stats to Statsd.

### CODE REVIEW

File: config/default.yaml [hunk 1]
```diff
@@ -65,4 +65,4 @@ feedbackListener:
     group: marathon-consumer-group
     sessionTimeout: 6000
     handleAllMessagesBeforeExiting: true
-    offsetResetStrategy: latest
+    offsetResetStrategy: latest
```
Comment (cscatolini) [inline config/default.yaml]: revert this change

File: migrations/20190325173135_add_status_and_events_table.sql [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+REFERENCES jobs(id)
+ON DELETE CASCADE
+ON UPDATE CASCADE;
+
+CREATE TYPE events_status AS ENUM ('fail', 'running', 'success');
+
+CREATE TABLE "events" (
+  "id" uuid DEFAULT uuid_generate_v4() UNIQUE,
+  "state" events_status NOT NULL,
+  "menssage" text NOT NULL,
```
Comment (cscatolini) [inline migrations/20190325173135_add_status_and_events_table.sql]: r/menssage/message

File: model/events.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+package model
+
+import (
+	"github.com/satori/go.uuid"
+)
+
+// Events is the events model struct
+type Events struct {
+	ID        uuid.UUID `sql:",pk" json:"id"`
+	Menssage  string    `json:"menssage"`
```
Comment (cscatolini) [inline model/events.go]: r/menssage/message

File: model/job.go [hunk 4]
```diff
@@ -98,3 +100,47 @@ func (j *Job) Validate(c echo.Context) error {
...
+		JobID:     j.ID,
+		ID:        uuid.NewV4(),
+		CreatedAt: time.Now().UnixNano(),
+	}
+	_, err := db.DB.Model(status).OnConflict("DO NOTHING").Insert()
+	if err != nil {
+		panic(err)
+	}
+	status = &Status{}
+	err = db.DB.Model(status).Where("job_id = ? AND name = ?", j.ID, name).First()
```
Comment (cscatolini) [inline model/job.go]: Why are you fetching the status you've just inserted into the db?

File: model/job.go [hunk 5]
```diff
@@ -98,3 +100,47 @@ func (j *Job) Validate(c echo.Context) error {
...
+	if err != nil {
+		panic(err)
+	}
+	event := &Events{
+		Menssage:  menssage,
```
Comment (cscatolini) [inline model/job.go]: r/menssage/message

File: model/job.go [hunk 6]
```diff
@@ -98,3 +100,47 @@ func (j *Job) Validate(c echo.Context) error {
...
+	}
+	_, err = db.DB.Model(event).Insert()
+	if err != nil {
+		panic(err)
+	}
+}
+
+// TagSuccess create a status in one job
+func (j *Job) TagSuccess(db *extensions.PGClient, name, menssage string) {
+	j.tag(db, name, menssage, "success")
```
Comment (cscatolini) [inline model/job.go]: r/menssage/message

File: model/job.go [hunk 7]
```diff
@@ -98,3 +100,47 @@ func (j *Job) Validate(c echo.Context) error {
...
+}
+
+// TagError create a status in one job
+func (j *Job) TagError(db *extensions.PGClient, name, menssage string) {
+	j.tag(db, name, menssage, "fail")
```
Comment (cscatolini) [inline model/job.go]: r/menssage/message

File: model/job.go [hunk 8]
```diff
@@ -98,3 +100,47 @@ func (j *Job) Validate(c echo.Context) error {
...
+}
+
+// TagRunning create a status in one job
+func (j *Job) TagRunning(db *extensions.PGClient, name, menssage string) {
+	j.tag(db, name, menssage, "running")
```
Comment (cscatolini) [inline model/job.go]: r/menssage/message

File: worker/create_batches_from_filters_worker.go [hunk 9]
```diff
@@ -53,6 +54,20 @@ type CreateBatchesFromFiltersWorker struct {
 	RedisClient               *redis.Client
 }
 
+// PrintMemUsage ...
+func PrintMemUsage() {
+	var m runtime.MemStats
+	runtime.ReadMemStats(&m)
+	// For info on each, see: https://golang.org/pkg/runtime/#MemStats
+	fmt.Printf("Alloc = %v MiB", bToMb(m.Alloc))
```
Comment (cscatolini) [inline worker/create_batches_from_filters_worker.go]: Even if this is for debug issues I'd rather use the logger instead of `fmt.Printf`.

File: worker/create_batches_from_filters_worker.go [hunk 10]
```diff
@@ -268,39 +291,65 @@ func (b *CreateBatchesFromFiltersWorker) Process(message *workers.Msg) {
...
 	l.Info("starting create_batches_using_filters_worker")
+
 	job := &model.Job{
 		ID: id,
 	}
+	job.TagRunning(b.MarathonDB, "create_batches_using_filters_worker", "starting")
 	err = b.MarathonDB.DB.Model(job).Column("job.*", "App").Where("job.id = ?", job.ID).Select()
-	checkErr(l, err)
+	if err != nil {
+		job.TagError(b.MarathonDB, "create_batches_using_filters_worker", err.Error())
```
Comment (cscatolini) [inline worker/create_batches_from_filters_worker.go]: `create_batches_using_filters_worker` is repeated several times, maybe create a variable?

File: worker/create_batches_worker.go [hunk 11]
```diff
@@ -393,25 +394,39 @@ func (b *CreateBatchesWorker) Process(message *workers.Msg) {
 	}
 	if len(job.CSVPath) > 0 {
 		err = b.createBatchesUsingCSV(job, isReexecution, dbPageSize)
-		checkErr(l, err)
+		if err != nil {
+			job.TagError(b.MarathonDB, "create_batches_worker", err.Error())
```
Comment (cscatolini) [inline worker/create_batches_worker.go]: `create_batches_worker` is repeated several times, maybe create a variable?

File: worker/process_batch_worker.go [hunk 12]
```diff
@@ -191,20 +191,25 @@ func (batchWorker *ProcessBatchWorker) updateJobBatchesInfo(jobID uuid.UUID) err
...
+
 		log.I(l, "Finished all batches")
+		job.TagSuccess(batchWorker.MarathonDB, "process_batche_worker", "Finished all batches")
 		job.CompletedAt = time.Now().UnixNano()
 		_, err = batchWorker.MarathonDB.DB.Model(&job).Column("completed_at").Update()
 		if err != nil {
 			return err
 		}
-		at := time.Now().Add(10 * time.Minute).UnixNano()
+		at := time.Now().Add(5 * time.Second).UnixNano()
```
Comment (cscatolini) [inline worker/process_batch_worker.go]: You shouldn't change this! I think you did this to have faster feedback while testing locally, to solve this we should parametrize the delay by environment.

File: worker/create_batches_from_filters_worker.go [hunk 13]
```diff
@@ -171,8 +186,10 @@ func (b *CreateBatchesFromFiltersWorker) preprocessPages(job *model.Job, stageSt
 
 func (b *CreateBatchesFromFiltersWorker) processPages(c <-chan DBPage, writeToCSVCH chan<- *[]User, job *model.Job, wg *sync.WaitGroup, wgCSV *sync.WaitGroup, stageStatus *StageStatus) {
 	for page := range c {
+		PrintMemUsage()
 		users := b.getPageFromDBWithFilters(job, page)
 		b.Logger.Info("got users from db", zap.Int("usersInBatch", len(*users)))
+		b.Workers.Statsd.Incr("processPages", []string{GetPushDBTableName(job.App.Name, job.Service)}, 1)
```
Comment (cscatolini) [inline worker/create_batches_from_filters_worker.go]: Statsd metrics should be in snake case, so "process_pages", "write_user_page_into_csv", etc

File: model/job.go (hunk 4 above)
Comment (capella) [reply model/job.go]: During the insertion one conflict can happen, in this case, the fetch action is needed to know the existing status id.

File: worker/create_batches_from_filters_worker.go (hunk 10 above)
Comment (capella) [reply worker/create_batches_from_filters_worker.go]: I create one checkErr method for the CreateBatchesFromFiltersWorker struct to don't repeat the error sentence.

File: worker/create_batches_worker.go (hunk 11 above)
Comment (capella) [reply worker/create_batches_worker.go]: I create one checkErr method for the CreateBatcheWorker struct to don't repeat the error sentence.

File: worker/create_batches_from_filters_worker.go (hunk 10 above)
Comment (cscatolini) [reply worker/create_batches_from_filters_worker.go]: We still have `create_batches_using_filters_worker` in a few places. Even with the checkErr method (nice catch by the way) I'd have a variable with this string.

File: worker/process_batch_worker.go [hunk 14]
```diff
@@ -191,20 +191,25 @@ func (batchWorker *ProcessBatchWorker) updateJobBatchesInfo(jobID uuid.UUID) err
...
+
 		log.I(l, "Finished all batches")
+		job.TagSuccess(batchWorker.MarathonDB, "process_batche_worker", "Finished all batches")
 		job.CompletedAt = time.Now().UnixNano()
 		_, err = batchWorker.MarathonDB.DB.Model(&job).Column("completed_at").Update()
 		if err != nil {
 			return err
 		}
-		at := time.Now().Add(10 * time.Minute).UnixNano()
+		at := time.Now().Add(batchWorker.Config.GetDuration("workers.processBatch.intervalToSendCompletedJob")).UnixNano()
```
Comment (cscatolini) [inline worker/process_batch_worker.go]: great! just add some default values in the config files
